### ISSUE TITLE
OAuth2 support

### ISSUE THREAD

**Issue by bufferoverflow** [fossology/fossology#515]
Would be great to have OAuth2 support, this would enable us to
- use standard authentication systems on internal deployments
- use a Mozilla or GitHub account to login on public FOSSology instance

This is related to  #193

**Comment by bobgob** [comment]
So what is involved?  common-auth.php (siteminder_check()) simply validates that a specific environment variable is set and returns its value (the fossology user id).  This is the technique used for Siteminder and PingIdentity.

**Comment by bufferoverflow** [comment]
Here is a nice library http://oauth2-client.thephpleague.com/ that supports OAuth2 for PHP.

**Comment by bufferoverflow** [comment]
Here is some sample code to achieve this:

``composer.json`` should contain this:
```json
{
    "require": {
        "league/oauth2-client": "^1.3"
    }
}
```
The example is using [gitlab as a oauth service provider](http://doc.gitlab.com/ce/integration/oauth_provider.html). To test this, do
* register the application [here](https://gitlab.com/profile/applications) 
* add the ApplicationId- and Secret-hash to the ``index.php`` file below
* run the sample by executing: ``php -S localhost:8000``
* point your browser to http://localhost:8000

```php
<?php
require __DIR__ . '/vendor/autoload.php';
$provider = new \League\OAuth2\Client\Provider\GenericProvider([
    'clientId'                => 'hash',
    'clientSecret'            => 'hash',
    'redirectUri'             => 'http://localhost:8000',
    'urlAuthorize'            => 'https://gitlab.com/oauth/authorize',
    'urlAccessToken'          => 'https://gitlab.com/oauth/token',
    'urlResourceOwnerDetails' => 'https://gitlab.com/api/v3/user'
]);

if (!isset($_GET['code'])) {
    $authorizationUrl = $provider->getAuthorizationUrl();
    $_SESSION['oauth2state'] = $provider->getState();
    header('Location: ' . $authorizationUrl. "&scope=api&response_type=code");
    exit;

} else {
    try {
        $accessToken = $provider->getAccessToken('authorization_code', [
            'code' => $_GET['code']
        ]);
        echo "<pre>";
        echo $accessToken->getToken() . "\n";
        echo $accessToken->getRefreshToken() . "\n";

        $resourceOwner = $provider->getResourceOwner($accessToken);

        var_export($resourceOwner->toArray());

        echo "\n";
        echo "name  " . $resourceOwner->toArray()['name']  . "\n";
        echo "email " . $resourceOwner->toArray()['email'] . "\n";
        echo "</pre>";
    } catch (\League\OAuth2\Client\Provider\Exception\IdentityProviderException $e) {
        exit($e->getMessage());
    }
}
?>
```

**Comment by bufferoverflow** [comment]
I think this would also help providing a public instance where people can Login with GitLab or GitHub accounts.

**Comment by suppadeliux** [comment]
Where can i find a doc to implement OAuth?
Thank you

**Comment by GMishx** [comment]
Hello @suppadeliux. It should be available with #2123
